### PR TITLE
Registration flow fixed

### LINKED ISSUE #21514: Can create a user without the registration flow finished properly (reopened #17644)
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

storage

### Describe the bug

Reopening #17644 as suggested on [Slack here](https://cloud-native.slack.com/archives/C056XU905S6/p1688643055350429?thread_ts=1688559314.400539&cid=C056XU905S6)

There was a PR (#19488) that stated to solve the issue, but it did not. It can be reproduced with the latest nightly at the moment.

I can create a User entity with the registration flow throwing an Exception. Not all elements are run in the registration flow in such cases, but the user still exists. Neither a REGISTER event is saved for the user.

### Version

latest nightly (but present since at least 19.x)

### Expected behavior

If the user entity is created then the registration flow should execute until the end successfully. Probably the whole flow should run in a single transaction.

### Actual behavior

The user entity is created, but an error page is shown to the user. The registration flow is not executed until the end.

### How to Reproduce?

A realm is required where the following is set:
 - registration enabled
 - forgot password enabled

To reproduce:
 - Open the login page
 - Click on the _register_ link
 - Click on the _back to the login_ link
 - Click on the _forgot password_ link
 - Enter a valid, existing e-mail address to the form. Click on the Submit button
 - Press the back button in the browser until you get back to the registration form (~3 times?)
   - a _page is expired_ error will be shown somewhere. Don't care about it, just press the back button until you get back to the login page
 - Fill out the registration form & submit your registration

An error page is shown, but the user is created in the database. From now on the freshly created user can ask for a reset password link, and log in after that with their newly created password.

### Anything else?

This is a reopen of #17644 as suggested on [Slack here](https://cloud-native.slack.com/archives/C056XU905S6/p1688643055350429?thread_ts=1688559314.400539&cid=C056XU905S6)

There was a PR (#19488) that stated to solve the issue, but it didoes not. It can be reproduced with the latest nightly at the moment.

I have a PR in #19556 which solves the issue, opened months ago.

### PR DESCRIPTION
**Pull request by vilmosnagy**
Closes #21514

Hi,

I've reported a bug in #17644. #19488 should have resolved it, but it seems to fail anyhow.

See my comment here: https://github.com/keycloak/keycloak/pull/19488#issuecomment-1497569865

@mposolda @hmlnarik I think we should implement a similar solution in the reset password flow to the idp linking flow's solution. As far as I remember in the IDP Linking flow the attributes of the newly created user is stored in an `AuthNote` - would you mind a solution, where:
 - instead of the `setUser` call in `ResetCredentialChooseUser:119` we store the user's ID (?) in an auth note
   - we should refactor some of the `ResetCredentialChooseUser::authenticate` because of this change
 - in the `ResetCredentialEmail` we should use the previously stored AuthNote instead of the `getUser` method

What do you think about my suggestion? If I'd implement such changes in this PR would you merge it?

Regards,

### CODE REVIEW

**Comment by vilmosnagy** [comment]
@mposolda @hmlnarik could you take a look at my changes pls? It should solve the bug I've reported, and all the tests pass.

**Comment by vilmosnagy** [comment]
@mposolda @hmlnarik I think this bug still exists in the upstream keycloak version. Could you give me a few hints how to get this fix merged? Or what direction should we look into to solve the bug?

**Comment by vilmosnagy** [comment]
@thomasdarimont thanks for the review, I've dropped the typo fixes from the method.

There is still one public API-related change: the Javadoc of the `FormContext::setUser` has a new sentence, indicating (requiring?) that the method should clear the pre-existing user if there was a user already in the context.

**Comment by mabartos** [comment]
@vilmosnagy Thanks for creating this PR! Could you please review if the failure in Base IT (5) job is related? Seems like it is.

**Comment by mabartos** [comment]
@vilmosnagy Thank you :)) 

I've seen some of our teammates were involved in similar tasks, so we'd need to get reviews from them. However, at this moment, we don't have the capacity in our teams due to solving tasks with higher priority. Thanks for understanding. 

@mposolda @hmlnarik @rmartinc Could you please check it? Thanks

**Comment by vilmosnagy** [comment]
@mabartos  thanks for the update. I'm a bit sad because this bug has been present since at least version 19.x and was first reported about three months ago. Moreover, it's quite easy for everyday users to trigger.

I'm more than willing to work on finding a solution for this issue, yet it seems to have been neglected and left unresolved :'(

**Comment by vilmosnagy** [comment]
as suggested in the [CNCF Slack](https://cloud-native.slack.com/archives/C056XU905S6/p1688643055350429?thread_ts=1688559314.400539&cid=C056XU905S6) by @abstractj , I've duplicated the issue #17644 into #21514, and rephrased the title of this PR to close the newly opened issue.

**Comment by abstractj** [comment]
> as suggested in the [CNCF Slack](https://cloud-native.slack.com/archives/C056XU905S6/p1688643055350429?thread_ts=1688559314.400539&cid=C056XU905S6) by @abstractj , I've duplicated the issue #17644 into #21514, and rephrased the title of this PR to close the newly opened issue.

@vilmosnagy to clarify, I never suggested duplicating issues on GitHub. Instead, it was recommended to create a new issue if the existing ones didn't cover the bug fixed in your pull-request.

**Comment by vilmosnagy** [comment]
> it was recommended to create a new issue if the existing ones didn't cover the bug fixed in your pull-request.

It's the same issue. As I mentioned multiple times previously (eg on github [here](https://github.com/keycloak/keycloak/pull/19488#issuecomment-1497569865), on PR #19488) PR #19488 _did not_ solve issue #17644. I can still reproduce the bug described in #17644 even months after PR #19488 has been merged (and that PR states that it solves the mentioned issue).

It did not solve it. I cannot be more clear than that, sorry. I try my best - probably there's a language barrier, but I cannot be more simple: _The bug described in #17644 is still present in the current nightly keycloak. The bug described in #17644 was never solved._

**Comment by ahus1** [comment]
Thank you for preparing this fix, @vilmosnagy. 

@sguilhen and I had a look and found that while your proposed fix prevents the error message being shown, it prevents the safety measures in `AuthenticationProcessor#setAutheticatedUser()` which check if the user has already been set. 

I pushed one more commit to this PR, which now adds the same check before creating the user. This way, when a user uses the back navigation and attempts a registration, they will see an error message, and no user would be created.

I ran the `UserStorageTest` tests locally and they passed for me. @sguilhen and I also did a manual test, and it looked good to us. I'm now waiting for the CI to complete. 

Please let me know your thoughts on this change before I go ahead and merge it. 

Thanks!

File: server-spi-private/src/main/java/org/keycloak/authentication/FormContext.java
Comment (vilmosnagy) [inline server-spi-private/src/main/java/org/keycloak/authentication/FormContext.java:70]: @ahus1  I think with your rework this should be removed as well, but I'm not a 100% sure.

```suggestion
```

**Comment by vilmosnagy** [comment]
@ahus1 thanks for the review & rework, it looks good to me.

I've added a small suggestion - my original change would have changed the `setUser` function a bit. But with your rework the javadoc change is not needed.

Thanks again :)

**Comment by ahus1** [comment]
@sguilhen - could you please add your review to this issue? Once it has been reviewed, I'd then squash and merge it. I might also ask Marek P. for a review as he's involve in the core things.

Thanks!

**Review by sguilhen — APPROVED** [review]
Approving as this reflects what @ahus1 and I discussed last week about the proper way to fix this.

**Comment by ahus1** [comment]
@mposolda - this is now ready to be merged IMHO, both @sguilhen and me looked into this. 
As this touches the "core" parts of Keycloak, I ask you or someone from your team for a review. 

Once your review is in, I will squash and merge it.

Thanks!

**Comment by ahus1** [comment]
@mposolda - I really like the change you did so the registration page doesn't show up any more. Closing this one, and continuing the discussion on the other PR.

**Pull request closed without merging**
